src/api.ts

```diff
@@ -0,0 +1,284 @@
+import { IConfigurationService } from "vs/platform/configuration/common/configuration";
+import { IWorkbenchLayoutService } from "vs/workbench/services/layout/browser/layoutService";
+import { ITelemetryService } from "vs/platform/telemetry/common/telemetry";
+import { IWorkspaceContextService } from "vs/platform/workspace/common/workspace";
+import { IEditorService } from "vs/workbench/services/editor/common/editorService";
+import { IThemeService } from "vs/platform/theme/common/themeService";
+import { IContextMenuService } from "vs/platform/contextview/browser/contextView";
+import { IViewletService } from "vs/workbench/services/viewlet/browser/viewlet";
+import { IEditorGroupsService } from "vs/workbench/services/editor/common/editorGroupsService";
+import { createCSSRule } from "vs/base/browser/dom";
+import { IDisposable } from "vs/base/common/lifecycle";
+
+/**
+ * Client-side implementation of VS Code's API.
+ * TODO: Views aren't quite working.
+ * TODO: Implement menu items for views (for item actions).
+ * TODO: File system provider doesn't work.
+ */
+export const vscodeApi = (serviceCollection: ServiceCollection): typeof vscode => {
+	const getService = <T>(id: ServiceIdentifier<T>): T => serviceCollection.get<T>(id) as T;
+	const commandService = getService(ICommandService);
+	const notificationService = getService(INotificationService);
+	const fileService = getService(IFileService);
+	const viewsRegistry = Registry.as<IViewsRegistry>(ViewsExtensions.ViewsRegistry);
+
+	// It would be nice to just export what VS Code creates but it looks to me
+	// that it assumes it's running in the extension host and wouldn't work here.
+	// It is probably possible to create an extension host that runs in the
+	// browser's main thread, but I'm not sure how much jank that would require.
+	// We could have a web worker host but we want DOM access.
+	return {
+		EventEmitter: Emitter,
+		TreeItemCollapsibleState: extHostTypes.TreeItemCollapsibleState,
+		FileSystemError: extHostTypes.FileSystemError,
+		FileType: FileType,
+		Uri: URI,
+		commands: {
+			executeCommand: (commandId: string, ...args: any[]): any => {
+				return commandService.executeCommand(commandId, ...args);
+			},
+			registerCommand: (id: string, command: () => void): any => {
+				return CommandsRegistry.registerCommand(id, command);
+			},
+		},
+		window: {
+			registerTreeDataProvider: (id: string, dataProvider: ITreeViewDataProvider): void => {
+				const view = viewsRegistry.getView(id);
+				if (view) {
+					(view as ITreeViewDescriptor).treeView.dataProvider = dataProvider;
+				}
+			},
+			showErrorMessage: (message: string): void => {
+				notificationService.error(message);
+			},
+		},
+		workspace: {
+			registerFileSystemProvider: (scheme: string, provider: vscode.FileSystemProvider): IDisposable => {
+				return fileService.registerProvider(scheme, new FileSystemProvider(provider));
+			},
+		},
+	} as any;
```

Comment:
Originally it was because it only implements a tiny part of the API (so it would get "X property is missing" errors) but I think what I'll do instead is return `Partial<typeof vscode>`.